chain/client/src/sync/state.rs

```diff
@@ -528,62 +621,167 @@ impl StateSync {
-            download.last_target = Some(make_account_or_peer_id_or_hash(target.clone()));
-            let run_me = download.run_me.clone();
-
-            near_performance_metrics::actix::spawn(
-                std::any::type_name::<Self>(),
-                self.network_adapter
-                    .send_async(PeerManagerMessageRequest::NetworkRequests(
-                        NetworkRequests::StateRequestPart {
-                            shard_id,
-                            sync_hash,
-                            part_id: part_id as u64,
-                            target: target.clone(),
-                        },
-                    ))
-                    .then(move |result| {
-                        // TODO: possible optimization - in the current code, even if one of the targets it not present in the network graph
-                        //       (so we keep getting RouteNotFound) - we'll still keep trying to assign parts to it.
-                        //       Fortunately only once every 60 seconds (timeout value).
-                        if let Ok(NetworkResponses::RouteNotFound) =
-                            result.map(|f| f.as_network_response())
-                        {
-                            // Send a StateRequestPart on the next iteration
-                            run_me.store(true, Ordering::SeqCst);
-                        }
-                        future::ready(())
-                    }),
-            );
+            match &self.mode {
+                StateSyncMode::Peers => {
+                    // For every part that needs to be requested it is selected one
+                    // peer (target) randomly to request the part from.
+                    // IMPORTANT: here we use 'zip' with possible_target_sampler -
+                    // which is limited. So at any moment we'll not request more
+                    // than possible_targets.len() * MAX_STATE_PART_REQUEST parts.
+                    let target = possible_targets_sampler.next().unwrap();
+                    self.request_part_from_peers(
+                        part_id as u64,
+                        target,
+                        download,
+                        shard_id,
+                        sync_hash,
+                    );
+                }
+                StateSyncMode::HeaderFromPeersAndPartsFromExternal { chain_id, bucket } => {
+                    self.request_part_from_external_storage(
+                        part_id as u64,
+                        download,
+                        shard_id,
+                        epoch_height,
+                        state_num_parts,
+                        &chain_id,
+                        bucket.clone(),
+                    );
+                }
+            }
+        }
+    }
+
+    /// Starts an asynchronous network request to external storage to obtain the given state part.
+    fn request_part_from_external_storage(
```

Comment:
I'm glad I made this work, but please let me know if you like another implementation better.
This implementation: Create a download task for each state part. Run at most X tasks concurrently across all shards.
Alternative implementation: Run as many async tasks as there are shards, where each task does the following:
* Check which state parts are missing from the RocksDB.
* Check which missing state parts are available in the S3 directory by doing `GetListing`.
* Sequentially download and write to DB state parts that are available on S3 but are not available locally. 
* If not done yet, then wait for the duration of `state_sync_timeout` and repeat.

The alternative solution will need an additional effort to speed up all those files faster than sequentially. If an S3 bucket is on the side of the world, the speed-of-light latency of network requests is at least 150ms, therefore it'll take at least 9000s to download 60k files, i.e. 2.5 hours of just waiting for the network latencies.